### PR TITLE
feat(es‑abstract): Update to v1.17

### PR DESCRIPTION
**Pull request by ExE-Boss**
~~Note that I haven’t yet extracted the methods into versioned sub‑directories.~~

It might be better to add **JSDoc** to `es‑abstract` directly instead.

## Depends on:
- [x] <https://github.com/DefinitelyTyped/DefinitelyTyped/pull/45117> (merged)
- [x] <https://github.com/DefinitelyTyped/DefinitelyTyped/pull/45118> (merged)
- [x] <https://github.com/DefinitelyTyped/DefinitelyTyped/pull/45231> (merged)
- [x] <https://github.com/DefinitelyTyped/DefinitelyTyped/pull/45326> (merged)
- [x] <https://github.com/DefinitelyTyped/DefinitelyTyped/pull/45333> (merged)
- [x] <https://github.com/DefinitelyTyped/DefinitelyTyped/pull/45337> (merged)
- [x] <https://github.com/DefinitelyTyped/DefinitelyTyped/pull/45641> (merged)

---

Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: <https://github.com/ljharb/es-abstract/blob/master/CHANGELOG.md>
- [x] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.
- [x] Include [tests for your changes](https://github.com/DefinitelyTyped/DefinitelyTyped#testing)
- [x] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`. If for reason the any rule need to be disabled, disable it for that line using `// tslint:disable-next-line [ruleName]` and not for whole package so that the need for disabling can be reviewed.

---

review?(@ljharb, @sandersn, @sheetalkamat)

### CODE REVIEW

**Review by ljharb — APPROVED** [review]

File: types/es-abstract/es2017.d.ts
Comment (danger-public) [inline types/es-abstract/es2017.d.ts:4]: - :no_entry_sign: The root .editorconfig style specifies spaces for whitespace. Please use spaces in new or changed types.

**Comment by danger-public** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Fails</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:no_entry_sign:</td>
      <td>

  **types/es-abstract/es2016.d.ts#L4** - The root .editorconfig style specifies spaces for whitespace. Please use spaces in new or changed types.
  </td>
    </tr>
  
<tr>
      <td>:no_entry_sign:</td>
      <td>

  **types/es-abstract/es2018.d.ts#L4** - The root .editorconfig style specifies spaces for whitespace. Please use spaces in new or changed types.
  </td>
    </tr>
  
<tr>
      <td>:no_entry_sign:</td>
      <td>

  **types/es-abstract/es2019.d.ts#L4** - The root .editorconfig style specifies spaces for whitespace. Please use spaces in new or changed types.
  </td>
    </tr>
  </tbody>
</table>




<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against 10a0f14400fd3803eba656a3527a267c6ff2de4f
</p>

File: types/es-abstract/test/helpers/index.ts
Comment (ljharb) [inline types/es-abstract/test/helpers/index.ts:1]: why do we need types for es-abstract's own test files? nobody should be importing them.

Comment (ExE-Boss) [reply types/es-abstract/test/helpers/index.ts:1]: These files exist so that I don’t have to edit `tsconfig.json` with every release to add new test files to the `files` array: <https://github.com/DefinitelyTyped/DefinitelyTyped/blob/06eb285a3042aabe88c28961cdddff1b83fdf15f/types/es-abstract/tsconfig.json#L15-L18>

Compare that to what’s on the `master` branch: <https://github.com/DefinitelyTyped/DefinitelyTyped/blob/d9f8c5c420b3dcaf7317be4c19b478800f38740d/types/es-abstract/tsconfig.json#L15-L51>

Comment (ljharb) [reply types/es-abstract/test/helpers/index.ts:1]: can't the files array have `test/**/*.js` or something?

Comment (ExE-Boss) [reply types/es-abstract/test/helpers/index.ts:1]: No, because [`files`](https://www.typescriptlang.org/v2/en/tsconfig#files) doesn’t use globs and `include` [is disallowed by `types‑publisher`](https://github.com/microsoft/types-publisher/blob/54bd71c9fed3c13fefed2284c3fad62bfac66bdd/src/lib/definition-parser.ts#L288-L290).

Comment (ljharb) [reply types/es-abstract/test/helpers/index.ts:1]: ah, sad, npm's `files` does use globs.

File: types/es-abstract/GetIntrinsic.d.ts
Comment (ljharb) [inline types/es-abstract/GetIntrinsic.d.ts:121]: why is anything `any`, when DT now requires a TS version that supports `unknown`?

Comment (ExE-Boss) [reply types/es-abstract/GetIntrinsic.d.ts:121]: Because of ergonomics of consuming these types from **JavaScript** code.

This is also why the `GetIntrinsic` fallback overload return type was changed to `any`.

Comment (ljharb) [reply types/es-abstract/GetIntrinsic.d.ts:121]: Can you elaborate? JS code doesn't need to care about types; and either way, if `any` passes but `unknown` doesn't, you have a bug.

Comment (ExE-Boss) [reply types/es-abstract/GetIntrinsic.d.ts:121]: Currently, `GetIntrinsic("%Reflect.apply%", true)` returns `unknown`, so it needs to be type cast using:
```js
const reflectApply = /** @type {typeof Reflect.apply | undefined} */ (GetIntrinsic("%Reflect.apply%", true));
```

---

By using `any`, one can do:
```js
/** @type {typeof Reflect.apply | undefined} */
const reflectApply = GetIntrinsic("%Reflect.apply%", true);
```

or even:
```js
const reflectApply = GetIntrinsic("%Reflect.apply%", true);
```

Which is what you get without typings.

Comment (ljharb) [reply types/es-abstract/GetIntrinsic.d.ts:121]: i can see why it'd be nice to have the typings overload on the literal string input, and return the proper type - but i'm still not understanding any `any` is better than `unknown`, and if you're using jsdoc in your JS, the whole point is to force the types to be correct.

Comment (ExE-Boss) [reply types/es-abstract/GetIntrinsic.d.ts:121]: I’ve implemented that in <https://github.com/DefinitelyTyped/DefinitelyTyped/pull/45117>

File: types/es-abstract/2015/CanonicalNumericIndexString.d.ts
Comment (ljharb) [inline types/es-abstract/2015/CanonicalNumericIndexString.d.ts:1]:
```suggestion
declare function CanonicalNumericIndexString(value: string): number | undefined;
```

File: types/es-abstract/2015/CreateHTML.d.ts
Comment (ljharb) [inline types/es-abstract/2015/CreateHTML.d.ts:1]:
```suggestion
type NonSymbolPrimitive = null | undefined | number | bigint | boolean | string;
declare function CreateHTML(
  string: unknown,
  tag: string,
  attribute: string,
  value?: NonSymbolPrimitive | { toString: (...args: any[]) => string }
): string;
```

File: types/es-abstract/2015/DefinePropertyOrThrow.d.ts
Comment (ljharb) [inline types/es-abstract/2015/DefinePropertyOrThrow.d.ts:3]: can you explain the `ThisType<any>`?

File: types/es-abstract/2015/Get.d.ts
Comment (ljharb) [inline types/es-abstract/2015/Get.d.ts:3]:
```suggestion
declare function Get<O extends object, P extends PropertyKey>(O: O, P: P): P extends keyof O ? O[P] : unknown;
```

File: types/es-abstract/2015/GetIterator.d.ts
Comment (ljharb) [inline types/es-abstract/2015/GetIterator.d.ts:6]:
```suggestion
declare function GetIterator<I extends Iterator<unknown, unknown, unknown>>(obj: { [Symbol.iterator](): I }): I;
declare function GetIterator<O, I extends Iterator<unknown, unknown, unknown>>(obj: O, method: (this: O) => I): I;
```

File: types/es-abstract/2015/GetV.d.ts
Comment (ljharb) [inline types/es-abstract/2015/GetV.d.ts:3]:
```suggestion
declare function GetV<O, P extends PropertyKey>(O: O, P: P): P extends keyof O ? O[P] : unknown;
```

File: types/es-abstract/2015/HasOwnProperty.d.ts
Comment (ljharb) [inline types/es-abstract/2015/HasOwnProperty.d.ts:3]: is there a way to have this assert that `P extends keyof O`, instead of just returning a boolean?

File: types/es-abstract/2015/ObjectCreate.d.ts
Comment (ljharb) [inline types/es-abstract/2015/ObjectCreate.d.ts:1]:
```suggestion
declare function ObjectCreate(proto: object | null, internalSlotsList?: readonly []): object extends proto;
```
?

File: types/es-abstract/2015/SameValueZero.d.ts
Comment (ljharb) [inline types/es-abstract/2015/SameValueZero.d.ts:1]: can this be a predicate that asserts that `x` and `y` are the same type?

File: types/es-abstract/2016/SameValueNonNumber.d.ts
Comment (ljharb) [inline types/es-abstract/2016/SameValueNonNumber.d.ts:9]:
```suggestion
    y: typeof x,
```

File: types/es-abstract/2018/IsStringPrefix.d.ts
Comment (ljharb) [inline types/es-abstract/2018/IsStringPrefix.d.ts:1]: is there a way in typescript to do something like `p extends q` as a predicate, so that `'f'` is seen as a prefix of `'foo'`?

File: types/es-abstract/5/CheckObjectCoercible.d.ts
Comment (ljharb) [inline types/es-abstract/5/CheckObjectCoercible.d.ts:1]:
```suggestion
declare function CheckObjectCoercible<T extends {}>(value: T, errorMessage?: string): NonNullable<T>;
```

File: types/es-abstract/5/FromPropertyDescriptor.d.ts
Comment (ljharb) [inline types/es-abstract/5/FromPropertyDescriptor.d.ts:3]: can you elaborate on what's happening here?  is `TypedPropertyDescriptor` a builtin thing?

File: types/es-abstract/5/SameValue.d.ts
Comment (ljharb) [inline types/es-abstract/5/SameValue.d.ts:1]: similarly, this should be a predicate that asserts that `x` and `y` have the same type

File: types/es-abstract/5/StrictEqualityComparison.d.ts
Comment (ljharb) [inline types/es-abstract/5/StrictEqualityComparison.d.ts:1]: can this be a predicate that asserts that x is the same type as y?

File: types/es-abstract/helpers/DefineOwnProperty.d.ts
Comment (ljharb) [inline types/es-abstract/helpers/DefineOwnProperty.d.ts:13]: can this instead be a union of all the ES* `SameValue` types? same on the other args.

**Review by ljharb — CHANGES_REQUESTED** [review]
In general, it'd be great to replace every instance of `any` with `unknown`.

whew, this is a lot of files (nice work!). i hope you autogenerated the ones that es-abstract autogenerates - if so, maybe we could indicate those in .gitattributes so they don't show up by default in the PR diffs?

File: types/es-abstract/5/FromPropertyDescriptor.d.ts
Comment (ExE-Boss) [reply types/es-abstract/5/FromPropertyDescriptor.d.ts:3]: `TypedPropertyDescriptor` is a **TypeScript** built‑in interface from `lib.es5.d.ts`.

File: types/es-abstract/2015/DefinePropertyOrThrow.d.ts
Comment (ExE-Boss) [reply types/es-abstract/2015/DefinePropertyOrThrow.d.ts:3]: That’s a **TypeScript** marker interface that tells the **TypeScript** compiler that the inferred `this` type of function object properties shouldn’t be the enclosing object (which in this case is `PropertyDescriptor`), but is instead the type parameter of `ThisType`.

File: types/es-abstract/2015/SameValueZero.d.ts
Comment (ExE-Boss) [reply types/es-abstract/2015/SameValueZero.d.ts:1]: Unfortunately, that’s not easy to do.

File: types/es-abstract/2018/IsStringPrefix.d.ts
Comment (ExE-Boss) [reply types/es-abstract/2018/IsStringPrefix.d.ts:1]: **TypeScript** doesn’t currently support this.

File: types/es-abstract/5/CheckObjectCoercible.d.ts
Comment (ExE-Boss) [reply types/es-abstract/5/CheckObjectCoercible.d.ts:1]: This returns `NonNullable<T>`, which correctly describes what this function does.

File: types/es-abstract/5/SameValue.d.ts
Comment (ExE-Boss) [reply types/es-abstract/5/SameValue.d.ts:1]: **TypeScript** doesn’t currently expose this functionality, which is why `Object.is` is defined as returning `boolean`.

File: types/es-abstract/5/StrictEqualityComparison.d.ts
Comment (ExE-Boss) [reply types/es-abstract/5/StrictEqualityComparison.d.ts:1]: **TypeScript** doesn’t currently provide a way to declare a function as being equivalent to `x === y`.

File: types/es-abstract/2015/CreateHTML.d.ts
Comment (ExE-Boss) [reply types/es-abstract/2015/CreateHTML.d.ts:1]: `unknown` is sufficiently close to that I’d say.

File: types/es-abstract/2015/ObjectCreate.d.ts
Comment (ExE-Boss) [reply types/es-abstract/2015/ObjectCreate.d.ts:1]: That’s broken syntax.

---

This currently returns `any` to mirror `Object.create`.

File: types/es-abstract/5/StrictEqualityComparison.d.ts
Comment (ljharb) [reply types/es-abstract/5/StrictEqualityComparison.d.ts:1]: not `===`, but `x is typeof y`?

File: types/es-abstract/5/SameValue.d.ts
Comment (ljharb) [reply types/es-abstract/5/SameValue.d.ts:1]: it can't assert that `x is typeof y`?

File: types/es-abstract/5/CheckObjectCoercible.d.ts
Comment (ljharb) [reply types/es-abstract/5/CheckObjectCoercible.d.ts:1]: so if i call `CheckObjectCoercible<null>`, TS will correctly error out because null isn't non-nullable?

File: types/es-abstract/2015/CreateHTML.d.ts
Comment (ljharb) [reply types/es-abstract/2015/CreateHTML.d.ts:4]: right, which is why i think it's important to be explicit. It shouldn't be possible to pass a typecheck when calling this with `value` as a Symbol.

File: types/es-abstract/2015/CanonicalNumericIndexString.d.ts
Comment (ljharb) [reply types/es-abstract/2015/CanonicalNumericIndexString.d.ts]: this doesn't seem resolved? it still says `value: unknown`

File: types/es-abstract/2015/CreateHTML.d.ts
Comment (ljharb) [reply types/es-abstract/2015/CreateHTML.d.ts:4]: (including this in my review; i still think it needs to change - "sufficiently close" is still "not 100% accurate")

File: types/es-abstract/2015/Get.d.ts
Comment (ljharb) [reply types/es-abstract/2015/Get.d.ts]: this one also still says `any`

File: types/es-abstract/2015/GetIterator.d.ts
Comment (ljharb) [reply types/es-abstract/2015/GetIterator.d.ts]: this also still has `any`

File: types/es-abstract/2015/GetV.d.ts
Comment (ljharb) [reply types/es-abstract/2015/GetV.d.ts]: also here

File: types/es-abstract/2015/HasOwnProperty.d.ts
Comment (ljharb) [reply types/es-abstract/2015/HasOwnProperty.d.ts]: this isn't resolved until there's at least an answer to the question - why not:
```suggest
declare function HasOwnProperty(O: object, P: PropertyKey): P is keyof O;
```

File: types/es-abstract/5/CheckObjectCoercible.d.ts
Comment (sandersn) [reply types/es-abstract/5/CheckObjectCoercible.d.ts:1]: Yep, `Type 'null' does not satisfy the constraint '{}'.`

**Review by ljharb — CHANGES_REQUESTED** [review]
i'm seeing a number of resolved comments that don't either fix the issue, or answer the question.

**Review by ljharb — CHANGES_REQUESTED** [review]
(see my immediately prior review for a few remaining items)

File: types/es-abstract/2015/CreateHTML.d.ts
Comment (ExE-Boss) [reply types/es-abstract/2015/CreateHTML.d.ts:4]: The biggest issue is that when `bigdecimal` gets added, the `NonSymbolPrimitive` type definition will need to be updated, whereas `unknown` will automatically include `bigdecimal`.

File: types/es-abstract/5/CheckObjectCoercible.d.ts
Comment (ljharb) [reply types/es-abstract/5/CheckObjectCoercible.d.ts:1]: @sandersn wait, *almost*? how is `unknown` different from `{} | null | undefined`??

File: types/es-abstract/2016/SameValueNonNumber.d.ts
Comment (ExE-Boss) [reply types/es-abstract/2016/SameValueNonNumber.d.ts:6]: That won’t work, since `typeof x` is `string | boolean | symbol | object | null | undefined` and not the type of whatever was passed as the first parameter.

File: types/es-abstract/2015/CreateHTML.d.ts
Comment (ljharb) [reply types/es-abstract/2015/CreateHTML.d.ts:4]: I think that's a reasonable compromise, especially for this library, which won't need to support bigdecimals until the ES edition it lands in (if it does).

File: types/es-abstract/2015/ObjectCreate.d.ts
Comment (ljharb) [reply types/es-abstract/2015/ObjectCreate.d.ts:1]: Just because TS's builtin types are subpar doesn't mean this library's should be. at the very least, `object` seems like a better choice than `any`.

File: types/es-abstract/2015/GetPrototypeFromConstructor.d.ts
Comment (ljharb) [inline types/es-abstract/2015/GetPrototypeFromConstructor.d.ts:10]:
```suggestion
): null | object;
```
(these are the only valid prototype values)

File: types/es-abstract/2015/ArrayCreate.d.ts
Comment (ljharb) [inline types/es-abstract/2015/ArrayCreate.d.ts:1]:
```suggestion
declare function ArrayCreate(length: number, proto?: object | null): unknown[];
```

**Review by ljharb — CHANGES_REQUESTED** [review]
LGTM after these last two changes

File: types/es-abstract/2015/GetPrototypeFromConstructor.d.ts
Comment (ljharb) [reply types/es-abstract/2015/GetPrototypeFromConstructor.d.ts]: This is marked as resolved but still isn’t updated or explained.

File: types/es-abstract/2015/ArrayCreate.d.ts
Comment (ljharb) [reply types/es-abstract/2015/ArrayCreate.d.ts]: same here

**Review by ljharb — CHANGES_REQUESTED** [review]

File: types/es-abstract/2015/GetPrototypeFromConstructor.d.ts
Comment (ExE-Boss) [reply types/es-abstract/2015/GetPrototypeFromConstructor.d.ts]: Because I’ve already fixed it in <https://github.com/DefinitelyTyped/DefinitelyTyped/pull/44805/commits/a366c979a63c8f4dd312bcf18d51d26903424587> 3 days ago.

Comment (ljharb) [reply types/es-abstract/2015/GetPrototypeFromConstructor.d.ts]: Did you forget to push that commit? If i click on it, it says "This commit does not belong to any branch on this repository.".

Comment (ljharb) [reply types/es-abstract/2015/GetPrototypeFromConstructor.d.ts]: nvm, perhaps it was rebased out, but at the time i left this comment it didn't seem to be included.

**Review by ljharb — APPROVED** [review]

**Pull request merged**
